Shared/Transforms/Variable.cs

```diff
@@ -0,0 +1,334 @@
+
+            return alphanumericPattern.IsMatch(name);
+        }
+
+        public string Key
+        {
+            get
+            {
+                return string.Format("${0}", m_name);
+            }
+        }
+
+        public string TextLayerKey
+        {
+            get
+            {
+                return string.Format("$({0})", m_name);
+            }
+        }
+
+        public string Value
+        {
+            get
+            {
+                return GetVariableValue();
+            }
+        }
+
+        public override string ToString()
+        {
+            return string.Format("${0}_{1}", m_name, GetVariableValue());
+        }
+
+        private string GetVariableValue()
+        {
+            string result = m_value.ToString();
+
+            var @types = new Dictionary<Type, Action> {
+                { typeof(int), () => { result = m_value.ToString(); } },
+                { typeof(double), () => {result = m_value.ToString(); } },
+                { typeof(float), () => { result = m_value.ToString(); } },
+                { typeof(long), () => { result = m_value.ToString(); } },
+                { typeof(decimal), () => { result = m_value.ToString(); } },
+                { typeof(Expression), () => { result = m_value.ToString(); } },
+                { typeof(string), () => { result = string.Format("!{0}!", m_value.ToString()); } }
+            };
+
+            @types[m_value.GetType()]();
+
+            return result;
+        }
+    }
+
+    /// <summary>
+    /// Represents expression object that can be used in user defined variables and conditional transformations.
+    /// </summary>
+    public class Expression
+    {
+        private string m_expression;
+        private int m_expr_length;
+        private ExpreAction m_lastAction = ExpreAction.Undefined;
+
+        private enum Expressions
+        {
+            Add,
+            Sub,
+            Mul,
+            Div,
+            Mod
+        }
+
+        private enum ExpreAction
+        {
+            Value,
+            Operation,
+            Undefined
+        }
+
+        private string GetExpressionKey(Expressions exp)
+        {
+            return exp.ToString().ToLower();
+        }
+
+        /// <summary>
+        /// Represent addition operation for expression.
+        /// </summary>
+        public Expression Add()
+        {
+            if (m_lastAction == ExpreAction.Operation || m_lastAction == ExpreAction.Undefined)
+                throw new InvalidOperationException("Operations can be used only after value defenition.");
+            m_expression += string.Format((m_expr_length > 0 ? "_" : string.Empty) + "{0}", GetExpressionKey(Expressions.Add));
+            m_expr_length++;
+            m_lastAction = ExpreAction.Operation;
+
+            return this;
+        }
+
+        /// <summary>
+        /// Represent subtraction operation for expression.
+        /// </summary>
+        public Expression Sub()
```

Comment:
Code duplication  of ADD, SUB, MUL, DIV, MOD (beside of the enumeration value used).
Please write the code once (use method that gets the enum value).

Reply:
All right